File: simulators/nats/nodejs/publisher.js
#!/usr/bin/env node
/**
 * NATS data simulator for vortex-server (Node.js).
 *
 * Mirrors simulators/nats/python/publisher.py — same subjects, same
 * payloads, same CLI flags — so users can pick whichever runtime is
 * already installed.
 *
 *   node publisher.js                                 # Core, rates.marketData
 *   node publisher.js --mode=jetstream                # JetStream, orders.*
 *   node publisher.js --rate-ms=100 --count=500       # ~10/s, stop after 500
 */

import { connect, StringCodec } from "nats";

// ----- CLI -------------------------------------------------------------------

function parseArgs(argv) {
  const opts = {
    url: "nats://localhost:4222",
    mode: "core",
    subject: "rates.marketData",
    stream: "ORDERS",
    subjectPrefix: "orders",
    rateMs: 200,
    count: 0,
    seed: null,
  };
  for (const arg of argv.slice(2)) {
    const [k, vRaw] = arg.split("=", 2);
    const v = vRaw ?? "";
    switch (k) {
      case "--url":             opts.url = v; break;
      case "--mode":            opts.mode = v; break;
      case "--subject":         opts.subject = v; break;
      case "--stream":          opts.stream = v; break;
      case "--subject-prefix":  opts.subjectPrefix = v; break;
      case "--rate-ms":         opts.rateMs = Number(v); break;
      case "--count":           opts.count = Number(v); break;
      case "--seed":            opts.seed = Number(v); break;
      case "--help":
      case "-h":
        console.log("Usage: node publisher.js [--url=...] [--mode=core|jetstream] " +
                    "[--subject=...] [--stream=...] [--subject-prefix=...] " +
                    "[--rate-ms=N] [--count=N] [--seed=N]");
        process.exit(0);
      default:
        console.error(`unknown argument: ${arg}`);
        process.exit(2);
    }
  }
  if (!["core", "jetstream"].includes(opts.mode)) {
    console.error(`--mode must be 'core' or 'jetstream'`);
    process.exit(2);
  }
  return opts;
}

// ----- deterministic PRNG (matches the Python --seed flag semantics) ---------

function makeRng(seed) {
  if (seed == null) return Math.random;
  // Mulberry32: tiny, well-mixed, fine for synthetic feeds.
  let s = seed | 0;
  return function () {
    s = (s + 0x6D2B79F5) | 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

// ----- payload makers --------------------------------------------------------

const RATES_SYMBOLS = ["EURUSD", "GBPUSD", "USDJPY", "AUDUSD", "USDCHF"];
const RATES_MARKETS = ["LON", "NYC", "TKO"];
const ORDER_SYMBOLS = ["AAPL", "MSFT", "GOOG", "AMZN", "NVDA", "META", "TSLA"];
const ORDER_SIDES   = ["BUY", "SELL"];
const ORDER_STATES  = ["NEW", "PARTIAL", "FILLED", "CANCELLED"];

const nowIso = () => new Date().toISOString();
const pick   = (rng, xs) => xs[Math.floor(rng() * xs.length)];
const round5 = (n) => Number(n.toFixed(5));
const round2 = (n) => Number(n.toFixed(2));

function makeRate(rng, prices) {
  const market = pick(rng, RATES_MARKETS);
  const sym    = pick(rng, RATES_SYMBOLS);
  let mid = prices.get(sym) ?? (0.8 + rng() * 199.2);
  mid = Math.max(0.0001, mid + (rng() - 0.5) * mid * 0.001);
  prices.set(sym, mid);
  const spread = mid * 0.0001;
  const bid = round5(mid - spread / 2);
  const ask = round5(mid + spread / 2);
  return {
    MarketId: market,
    Id: sym,
    Bid: [bid, round5(bid - spread), round5(bid - spread * 2)],
    Ask: [ask, round5(ask + spread), round5(ask + spread * 2)],
    Timestamp: nowIso(),
  };
}

function makeOrder(rng, seq, prices) {
  const sym = pick(rng, ORDER_SYMBOLS);
  let mid = prices.get(sym) ?? (50 + rng() * 450);
  mid = Math.max(0.01, mid + (rng() - 0.5) * 0.5);
  prices.set(sym, mid);
  return {
    OrderId:   `ORD-${String(seq).padStart(8, "0")}`,
    Symbol:    sym,
    Side:      pick(rng, ORDER_SIDES),
    Qty:       1 + Math.floor(rng() * 1000),
    Price:     round2(mid),
    Status:    pick(rng, ORDER_STATES),
    Timestamp: nowIso(),
  };
}

// ----- publish loops ---------------------------------------------------------

async function publishCore(opts, rng) {
  const nc = await connect({ servers: opts.url, name: "vortex-sim-nats-core" });
  const sc = StringCodec();
  console.log(`[nats-core] connected to ${opts.url}, publishing to '${opts.subject}' ` +
              `at ~${(1000 / opts.rateMs).toFixed(1)} msg/s`);

  const prices = new Map();
  let sent = 0;
  const start = process.hrtime.bigint();
  installShutdown(async () => { await nc.flush(); await nc.close(); });

  while (opts.count === 0 || sent < opts.count) {
    nc.publish(opts.subject, sc.encode(JSON.stringify(makeRate(rng, prices))));
    sent++;
    if (sent % 50 === 0) {
      const elapsed = Number(process.hrtime.bigint() - start) / 1e9;
      console.log(`[nats-core] sent=${sent} actual_rate=${(sent / elapsed).toFixed(1)}/s`);
    }
    await sleep(opts.rateMs);
  }
  await nc.flush();
  await nc.close();
  console.log(`[nats-core] disconnected after ${sent} messages`);
}

async function publishJetstream(opts, rng) {
  const nc  = await connect({ servers: opts.url, name: "vortex-sim-nats-jetstream" });
  const jsm = await nc.jetstreamManager();
  const js  = nc.jetstream();
  const sc  = StringCodec();

  try {
    await jsm.streams.add({ name: opts.stream, subjects: [`${opts.subjectPrefix}.>`] });
    console.log(`[nats-js] created stream '${opts.stream}' (subjects: ${opts.subjectPrefix}.>)`);
  } catch (err) {
    // Stream already exists with compatible config — nothing to do.
    console.log(`[nats-js] stream '${opts.stream}' already exists`);
  }

  console.log(`[nats-js] connected to ${opts.url}, publishing under ` +
              `'${opts.subjectPrefix}.<symbol>' at ~${(1000 / opts.rateMs).toFixed(1)} msg/s`);

  const prices = new Map();
  let sent = 0;
  const start = process.hrtime.bigint();
  installShutdown(async () => { await nc.flush(); await nc.close(); });

  while (opts.count === 0 || sent < opts.count) {
    const row = makeOrder(rng, sent + 1, prices);
    const subject = `${opts.subjectPrefix}.${row.Symbol.toLowerCase()}`;
    await js.publish(subject, sc.encode(JSON.stringify(row)));
    sent++;
    if (sent % 50 === 0) {
      const elapsed = Number(process.hrtime.bigint() - start) / 1e9;
      console.log(`[nats-js] sent=${sent} actual_rate=${(sent / elapsed).toFixed(1)}/s`);
    }
    await sleep(opts.rateMs);
  }
  await nc.flush();
  await nc.close();
  console.log(`[nats-js] disconnected after ${sent} messages`);
}

// ----- plumbing --------------------------------------------------------------

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function installShutdown(closer) {
  let already = false;
  const handler = async () => {
    if (already) return;
    already = true;
    try { await closer(); } catch {}
    process.exit(130);
  };
  process.on("SIGINT",  handler);
  process.on("SIGTERM", handler);
}

(async () => {
  const opts = parseArgs(process.argv);
  const rng  = makeRng(opts.seed);
  try {
    if (opts.mode === "jetstream") await publishJetstream(opts, rng);
    else                            await publishCore(opts, rng);
  } catch (err) {
    console.error(`[fatal] ${err.message ?? err}`);
    process.exit(1);
  }
})();

File: simulators/nats/python/publisher.py
"""NATS data simulator for vortex-server.

Publishes either:
  - NATS Core messages to `rates.marketData`  (vortex table: RatesMarketData)
  - JetStream messages to `orders.*`          (vortex table: Orders)

The schema in each mode matches the corresponding entry in
`config.example.json` so the published rows are immediately usable by the
Perspective viewer once vortex-server is running.
"""

from __future__ import annotations

import argparse
import asyncio
import json
import random
import signal
import string
import sys
import time
from datetime import datetime, timezone

import nats
from nats.js.api import StreamConfig
from nats.js.errors import BadRequestError


# A handful of fake symbols so the data viewed in Perspective looks
# realistic and has interesting grouping behaviour.
RATES_SYMBOLS = ["EURUSD", "GBPUSD", "USDJPY", "AUDUSD", "USDCHF"]
RATES_MARKETS = ["LON", "NYC", "TKO"]
ORDER_SYMBOLS = ["AAPL", "MSFT", "GOOG", "AMZN", "NVDA", "META", "TSLA"]
ORDER_SIDES   = ["BUY", "SELL"]
ORDER_STATES  = ["NEW", "PARTIAL", "FILLED", "CANCELLED"]


def now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")


def make_rate(prices: dict[str, float]) -> dict:
    """Build a RatesMarketData row with nested Bid/Ask price ladders."""
    market = random.choice(RATES_MARKETS)
    sym = random.choice(RATES_SYMBOLS)
    mid = prices.setdefault(sym, random.uniform(0.8, 200.0))
    # Random-walk the mid price.
    mid = max(0.0001, mid + random.uniform(-mid * 0.0005, mid * 0.0005))
    prices[sym] = mid
    spread = mid * 0.0001
    bid = round(mid - spread / 2, 5)
    ask = round(mid + spread / 2, 5)
    # Three-level ladder, slightly worse on each level.
    bids = [round(bid - i * spread, 5) for i in range(3)]
    asks = [round(ask + i * spread, 5) for i in range(3)]
    return {
        "MarketId": market,
        "Id": sym,
        "Bid": bids,
        "Ask": asks,
        "Timestamp": now_iso(),
    }


def make_order(seq: int, prices: dict[str, float]) -> dict:
    sym = random.choice(ORDER_SYMBOLS)
    mid = prices.setdefault(sym, random.uniform(50.0, 500.0))
    mid = max(0.01, mid + random.uniform(-0.25, 0.25))
    prices[sym] = mid
    return {
        "OrderId": f"ORD-{seq:08d}",
        "Symbol": sym,
        "Side": random.choice(ORDER_SIDES),
        "Qty": random.randint(1, 1000),
        "Price": round(mid, 2),
        "Status": random.choice(ORDER_STATES),
        "Timestamp": now_iso(),
    }


async def publish_core(args: argparse.Namespace) -> None:
    """Plain NATS Core publish loop. Subject must match vortex-server's
    `transports.nats` + `tables[*].source.subject` for the target table."""
    nc = await nats.connect(args.url, name="vortex-sim-nats-core")
    print(f"[nats-core] connected to {args.url}, "
          f"publishing to '{args.subject}' at ~{1000 / args.rate_ms:.1f} msg/s",
          flush=True)
    prices: dict[str, float] = {}
    sent = 0
    start = time.monotonic()
    try:
        while args.count == 0 or sent < args.count:
            row = make_rate(prices)
            await nc.publish(args.subject, json.dumps(row).encode())
            sent += 1
            if sent % 50 == 0:
                rate = sent / max(time.monotonic() - start, 1e-6)
                print(f"[nats-core] sent={sent} actual_rate={rate:.1f}/s", flush=True)
            await asyncio.sleep(args.rate_ms / 1000.0)
    finally:
        await nc.flush()
        await nc.close()
        print(f"[nats-core] disconnected after {sent} messages", flush=True)


async def publish_jetstream(args: argparse.Namespace) -> None:
    """JetStream publish loop. Ensures the target stream exists (idempotent)
    so this script can bootstrap a brand-new NATS server without manual setup."""
    nc = await nats.connect(args.url, name="vortex-sim-nats-jetstream")
    js = nc.jetstream()
    try:
        await js.add_stream(StreamConfig(name=args.stream, subjects=[f"{args.subject_prefix}.>"]))
        print(f"[nats-js] created stream '{args.stream}' "
              f"(subjects: {args.subject_prefix}.>)", flush=True)
    except BadRequestError:
        # Stream already exists with compatible config — nothing to do.
        print(f"[nats-js] stream '{args.stream}' already exists", flush=True)

    print(f"[nats-js] connected to {args.url}, "
          f"publishing under '{args.subject_prefix}.<symbol>' at "
          f"~{1000 / args.rate_ms:.1f} msg/s", flush=True)

    prices: dict[str, float] = {}
    sent = 0
    start = time.monotonic()
    try:
        while args.count == 0 or sent < args.count:
            row = make_order(sent + 1, prices)
            subject = f"{args.subject_prefix}.{row['Symbol'].lower()}"
            await js.publish(subject, json.dumps(row).encode())
            sent += 1
            if sent % 50 == 0:
                rate = sent / max(time.monotonic() - start, 1e-6)
                print(f"[nats-js] sent={sent} actual_rate={rate:.1f}/s", flush=True)
            await asyncio.sleep(args.rate_ms / 1000.0)
    finally:
        await nc.flush()
        await nc.close()
        print(f"[nats-js] disconnected after {sent} messages", flush=True)


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--url",      default="nats://localhost:4222",
                   help="NATS connection URL")
    p.add_argument("--mode",     choices=("core", "jetstream"), default="core",
                   help="Publish via Core pub/sub or via JetStream")
    p.add_argument("--subject",  default="rates.marketData",
                   help="Core mode: subject to publish to")
    p.add_argument("--stream",   default="ORDERS",
                   help="JetStream mode: stream name (created if missing)")
    p.add_argument("--subject-prefix", default="orders",
                   help="JetStream mode: subject prefix (full subject = <prefix>.<symbol>)")
    p.add_argument("--rate-ms",  type=int, default=200,
                   help="Inter-message delay (default 200ms ≈ 5 msg/s)")
    p.add_argument("--count",    type=int, default=0,
                   help="Stop after this many messages (0 = forever)")
    p.add_argument("--seed",     type=int, default=None,
                   help="Optional RNG seed for reproducible payloads")
    return p.parse_args()


def install_sigint_handler(loop: asyncio.AbstractEventLoop) -> None:
    """Convert SIGINT/SIGTERM to a graceful cancel of the current task."""
    def cancel():
        for task in asyncio.all_tasks(loop):
            task.cancel()
    for sig in (signal.SIGINT, signal.SIGTERM):
        try:
            loop.add_signal_handler(sig, cancel)
        except NotImplementedError:
            # Windows doesn't support add_signal_handler in proactor loop.
            pass


def main() -> int:
    args = parse_args()
    if args.seed is not None:
        random.seed(args.seed)

    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    install_sigint_handler(loop)

    coro = publish_jetstream(args) if args.mode == "jetstream" else publish_core(args)
    try:
        loop.run_until_complete(coro)
        return 0
    except (KeyboardInterrupt, asyncio.CancelledError):
        print("\n[interrupted]", flush=True)
        return 130
    finally:
        loop.close()


if __name__ == "__main__":
    sys.exit(main())
